File: 2021/2021W19/PreppindataW19.py
import pandas as pd
import re

xlsx = pd.ExcelFile("F:/Data/PD 2021 Week 19 Input.xlsx")

lookup = {}
for sheet in [sh for sh in xlsx.sheet_names if re.match('.*Lookup Table$',sh)]:
    df = pd.read_excel(xlsx,sheet)
    col = df.columns[1]
    df.columns = ['Code', 'Name']
    df['Code'] = df['Code'].str.lower()
    lookup[col] = {k:v[0] for (k,v) in df.set_index('Code').T.to_dict('list').items()}

sched =  pd.read_excel(xlsx,'Project Schedule Updates')
sched['Week'] = sched['Week'].apply(lambda x: f'Week {x}')
sched['Commentary'] = sched['Commentary'].apply(lambda x: re.sub('(\s+)(?=\[)','@',x))
sched = pd.concat([sched['Week'], 
                   pd.DataFrame([map(str.strip, x) \
                                 for x in sched['Commentary'].str.split('@').values.tolist()])],
                  axis=1, sort=False)
sched = sched.melt(id_vars='Week', value_name='Commentary', var_name='ToDrop').drop('ToDrop', axis=1).dropna().reset_index(drop=True)
sched['Detail'] = sched['Commentary'].apply(lambda x: re.search('(?<=\]\s)(.*)',x).group(1))
sched['Days Noted'] = sched['Commentary'].apply(lambda x: int(re.search('(\d+)(?=\sday)',x).group(1)) if re.search('(\d+)(?=\sday)',x) else None)
for c in [('Name','(\w+)(?=\.$)'),('Project','(?<=\[)(\w+)'),('Sub-Project','(?<=/)(Mar|Op)'),('Task','(\w+)(?=\])')]:
    sched[c[0]] = sched['Commentary'].apply(lambda x: lookup[c[0]][(re.search(c[1],x).group(1)).lower()])
sched = sched[['Week', 'Project', 'Sub-Project', 'Task', 'Name', 'Days Noted', 'Detail']]    

File: 2021/2021W19/preppindataW19.R
library(readxl)
library(dplyr)
library(stringr)
library(splitstackshape)

xlsx <- "F:/Data/PD 2021 Week 19 Input.xlsx"
final <- read_excel(xlsx,'Project Schedule Updates') %>%
  mutate('Week' = paste('Week',`Week`),
         'Commentary' = str_replace_all(`Commentary`,'(\\s+)(?=\\[)','@')) %>%
  cSplit(., 'Commentary', '@') %>%
  pivot_longer(cols=-'Week', names_to='toDrop', values_to = 'Commentary', values_drop_na=TRUE) %>%
  mutate('Project Code' = str_extract(`Commentary`,'(?<=\\[)(\\w+)'),
         'Sub-Project Code' = str_to_lower(str_extract(`Commentary`,'(?<=/)(Mar|Op)')),
         'Task Code' = str_to_title(str_extract(`Commentary`,'(\\w+)(?=\\])')),
         'Abbreviation' = str_to_title(str_extract(`Commentary`,'(\\w+)(?=\\.$)')),
         'Days Noted' = as.integer(str_extract(`Commentary`,'(\\d+)(?=\\sday)')),
         'Detail' = str_extract(`Commentary`,'(?<=\\]\\s)(.*)')) %>%
  merge(., read_excel(xlsx,'Project Lookup Table'), on='Project Code') %>%
  merge(., read_excel(xlsx,'Sub-Project Lookup Table'), on='Sub-Project Code') %>%
  merge(., read_excel(xlsx,'Task Lookup Table'), on='Task Code') %>%
  merge(., read_excel(xlsx,'Owner Lookup Table'), on='Abbreviation') %>%
  select(c('Week', 'Project', 'Sub-Project', 'Task', 'Name', 'Days Noted', 'Detail'))

View(final)
